README.md
# ScanPort
The program is a solution to the task from the second laboratory on the subject "Network Information Technologies".

## Overview
This program is a port scanner. Its main functionality:
- Full port scan from 0 to 49151;
- Scanning ports for a pattern that the user specifies (for example, 1,80,21,1000-1010);
- 2 operating modes: order | speed;
- Ability to save the last scan and display it;
- All reference information and documentation are described in script files;
- And many other important things.

## LICENSE
The full text of the license can be found at the following [link](https://github.com/Nakama3942/ScanPort/blob/main/LICENSE).

> Copyright © 2022 Kalynovsky Valentin. All rights reserved.
> 
> Licensed under the Apache License, Version 2.0 (the "License");
> you may not use this file except in compliance with the License.
> You may obtain a copy of the License at
> 
>     http://www.apache.org/licenses/LICENSE-2.0
> 
> Unless required by applicable law or agreed to in writing, software
> distributed under the License is distributed on an "AS IS" BASIS,
> WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> See the License for the specific language governing permissions and
> limitations under the License.

## Usage
### With Python
To start the program, just run *main.pyw*.

But it is also possible to run the program through the console:
```shell
python main.pyw
```
If the program gives a missing modules' error at startup, install Qt:
```shell
pip install PyQt6
# or
pip install pyqt6
```
### Without Python
There is a compiled version of the program. To use it, simply download the archive in the binaries' section under the latest version release. Unzip the archive and then use it as a regular program.


## Authors
<table align="center" style="border-width: 10; border-style: ridge">
	<tr>
		<td align="center"><a href="https://github.com/Nakama3942"><img src="https://avatars.githubusercontent.com/u/73797846?s=400&u=a9b7688ac521d739825d7003a5bd599aab74cb76&v=4" width="150px;" alt=""/><br /><sub><b>Kalynovsky Valentin</b></sub></a><sub><br />"Ideological inspirer and Author"</sub></td>
		<!--<td></td>-->
	</tr>
<!--
	<tr>
		<td></td>
		<td></td>
	</tr>
-->
</table>

CHANGELOG.md
# Changelog
<!--
Copyright © 2022 Kalynovsky Valentin. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->
<!--
## vX.X.X (DATE)

#### Bug Fixes:
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Invalid Fixed:
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Documenting:
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Duplicating:
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Enhancements:
- [# XXX](https : / / github . com / XXX) DESCRIPTION

---
-->
## v1.0.1 (15.10.2022)

#### Documenting:
- Added the CHANGELOG.md
- Updated the README.md

#### Enhancements:
- Added program icon;
- main.py renamed to main.pyw to run the program without a console;
- project compilation implemented.

---

## v1.0.0 (10.10.2022)

#### Release
This program is a port scanner. Its main functionality:
- Full port scan from 0 to 49151;
- Scanning ports for a pattern that the user specifies (for example, 1,80,21,1000-1010);
- 2 operating modes: order | speed;
- Ability to save the last scan and display it;
- All reference information and documentation are described in script files;
- And many other important things.


ui/scanport.py
#  Copyright © 2022 Kalynovsky Valentin. All rights reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import socket
from threading import Thread
from time import sleep
import pickle
import os

from PyQt6 import QtWidgets
from PyQt6.QtWidgets import QApplication, QMainWindow, QMessageBox
from PyQt6.QtCore import QRegularExpression, QDir
from PyQt6.QtGui import QRegularExpressionValidator, QIcon
from PyQt6.QtCore import QThread, pyqtSignal

from ui.raw.ui_scanport import Ui_ScanPort


# This class is a QThread that runs a function that tests a port to see if it's open
class PortTestThread(QThread):
    attempt = pyqtSignal(bool, str, int)

    def __init__(self, host: str, port: int):
        super(PortTestThread, self).__init__()
        self.host: str = host
        self.port: int = port

    def run(self):
        """
        It attempts to connect to a host on a given port, and it emits a signal with result to connect, the host and port
        number
        """
        s = socket.socket()
        try:
            s.connect((self.host, self.port))  # Connection attempt
            # s.settimeout(0.2)  # Timeout for a little more speed
        except:
            self.attempt.emit(False, self.host, self.port)  # Port is closed
        else:
            self.attempt.emit(True, self.host, self.port)  # Port is opened


class ScanPort(QMainWindow, Ui_ScanPort):
    def __init__(self):
        super(ScanPort, self).__init__()
        self.setupUi(self)

        # Icon initialization
        QDir.addSearchPath('icons', 'icons/')
        self.setWindowIcon(QIcon('icons:ScanPort_Icon.png'))

        # Data
        self.FLAG_STOP: bool = False
        self.port_list: list[list] = []

        # It's creating a validator for the input field of the list of ports
        total = '(\*)'
        number = '(\d|[1-9]\d{1,3}|[1-3]\d{4}|4[0-8]\d{3}|490\d{2}|491[0-4]\d|4915[01])'
        diapason = '(-)'
        separator = '(,)'
        volume = '{0,}'
        rx = QRegularExpression(f'^{total}|({number}{diapason}{number}{separator}|{number}{separator}){volume}$')
        validator = QRegularExpressionValidator(rx, self)
        self.linePorts.setValidator(validator)

        # It's a tracking of button clicks in the window
        self.buttScan.clicked.connect(self.buttScan_Clicked)
        self.linePorts.textChanged.connect(self.buttScan_Active)
        self.toolClear.clicked.connect(self.toolClear_Clicked)
        self.buttSave.clicked.connect(self.buttSave_Clicked)
        self.buttOpen.clicked.connect(self.buttOpen_Clicked)
        self.toolDelete.clicked.connect(self.toolDelete_Clicked)
        self.toolStop.clicked.connect(self.toolStop_Clicked)

        # It's checking if there is a file with the last saved data, and if there is, it enables the button to open it
        try:
            with open("data/LastPortsTest.save", "rb"):
                self.buttOpen.setEnabled(True)
                self.toolDelete.setEnabled(True)
        except IOError:
            pass

    def buttScan_Active(self):
        """
        If the text in the linePorts is not empty, enable the buttScan button
        """
        if self.linePorts.text() != '':
            self.buttScan.setEnabled(True)
        else:
            self.buttScan.setEnabled(False)

    def buttScan_Clicked(self):
        """
        It starts a thread that runs the portscan function
        """
        update = Thread(target=self.portscan, name='Tester', daemon=True)
        update.start()

    def portscan(self):
        """
        Gets ports and scans them
        """
        self.lineHost.setEnabled(False)
        self.linePorts.setEnabled(False)
        self.buttScan.setEnabled(False)
        self.toolClear.setEnabled(False)
        self.buttSave.setEnabled(False)
        self.buttOpen.setEnabled(False)
        self.toolStop.setEnabled(True)

        if self.lineHost.text() == '':
            self.lineHost.setText(self.lineHost.placeholderText())
        host = self.lineHost.text()

        if self.linePorts.text() == '*':
            for item in range(0, 49151):
                if self.FLAG_STOP:
                    self.FLAG_STOP = False
                    break
                update_thread = PortTestThread(host, item)
                update_thread.start()
                update_thread.attempt.connect(self.publish_result)
                if self.radioOrder.isChecked():
                    update_thread.wait()
                sleep(0.1)
        else:
            if self.linePorts.text()[-1] == ',' or self.linePorts.text()[-1] == '-':
                self.linePorts.setText(self.linePorts.text()[:-1])

            self.port_list = self.linePorts.text().split(',')
            for i in range(len(self.port_list)):
                self.port_list[i] = self.port_list[i].split('-')

            for item in self.port_list:
                if self.FLAG_STOP:
                    self.FLAG_STOP = False
                    break
                if len(item) == 1:
                    update_thread = PortTestThread(host, int(item[0]))
                    update_thread.start()
                    update_thread.attempt.connect(self.publish_result)
                    if self.radioOrder.isChecked():
                        update_thread.wait()
                    sleep(0.1)
                elif len(item) == 2:
                    if int(item[0]) < int(item[1]):
                        for port in range(int(item[0]), int(item[1]) + 1, 1):
                            if self.FLAG_STOP:
                                break
                            update_thread = PortTestThread(host, port)
                            update_thread.start()
                            update_thread.attempt.connect(self.publish_result)
                            if self.radioOrder.isChecked():
                                update_thread.wait()
                            sleep(0.1)
                    else:
                        for port in range(int(item[0]), int(item[1]) - 1, -1):
                            if self.FLAG_STOP:
                                break
                            update_thread = PortTestThread(host, port)
                            update_thread.start()
                            update_thread.attempt.connect(self.publish_result)
                            if self.radioOrder.isChecked():
                                update_thread.wait()
                            sleep(0.1)

        self.lineHost.setEnabled(True)
        self.linePorts.setEnabled(True)
        self.buttScan.setEnabled(True)
        self.toolClear.setEnabled(True)
        self.buttSave.setEnabled(True)
        try:
            with open("data/LastPortsTest.save", "rb"):
                self.buttOpen.setEnabled(True)
        except IOError:
            self.buttOpen.setEnabled(False)
        self.toolStop.setEnabled(False)

    def publish_result(self, result: bool, host: str, port: int):
        """
        A function that is used to publish the result of the port scan

        :param result: bool
        :type result: bool
        :param host: The hostname to connect to
        :type host: str
        :param port: The port to scan
        :type port: int
        """
        # My example for scan: 10,21,50,100,1000-1100,45155,80
        if result:
            self.textResult.append(f"<span style='color: darkgreen;'>[+] {host} : {port} opened</span>")
        else:
            self.textResult.append(f"<span style='color: darkred;'>[!] {host} : {port} closed</span>")

    def toolClear_Clicked(self):
        """
        It clears the text in the lineHost, linePorts, and textResult widgets, disables the toolClear and buttSave buttons,
        and enables the buttOpen button if the file "data/LastPortsTest.save" exists
        """
        self.lineHost.clear()
        self.linePorts.clear()
        self.textResult.clear()
        self.toolClear.setEnabled(False)
        self.buttSave.setEnabled(False)
        try:
            with open("data/LastPortsTest.save", "rb"):
                self.buttOpen.setEnabled(True)
        except IOError:
            self.buttOpen.setEnabled(False)

    def buttSave_Clicked(self):
        """
        It saves the data from the GUI to a file
        """
        data = [str(self.lineHost.text()), str(self.linePorts.text()), str(self.textResult.toHtml())]
        if not os.path.exists('data'):
            os.makedirs('data')
        with open("data/LastPortsTest.save", "wb") as save:
            pickle.dump(data, save)
        self.buttSave.setEnabled(False)
        self.buttOpen.setEnabled(False)
        self.toolDelete.setEnabled(True)
        inform = QMessageBox()
        inform.setText("Save done")
        inform.setIcon(QMessageBox.Icon.Information)
        inform.setStandardButtons(QMessageBox.StandardButton.Ok)
        ret: int = inform.exec()
        match ret:
            case QMessageBox.StandardButton.Ok:
                return

    def buttOpen_Clicked(self):
        """
        It opens a file, loads the data, and sets the data to the text boxes
        """
        with open("data/LastPortsTest.save", "rb") as save:
            data = pickle.load(save)
            self.lineHost.setText(data[0])
            self.linePorts.setText(data[1])
            self.textResult.setText(data[2])
        self.toolClear.setEnabled(True)
        self.buttSave.setEnabled(False)
        self.buttOpen.setEnabled(False)

    def toolDelete_Clicked(self):
        """
        It deletes a file and disables two buttons
        """
        os.remove("data/LastPortsTest.save")
        self.buttOpen.setEnabled(False)
        self.toolDelete.setEnabled(False)

    def toolStop_Clicked(self):
        """
        The function is called when the user clicks on the "Stop" button
        """
        self.FLAG_STOP = True


ui/raw/ui_scanport.py
# Form implementation generated from reading ui file 'ScanPort.ui'
#
# Created by: PyQt6 UI code generator 6.3.1
#
# WARNING: Any manual changes made to this file will be lost when pyuic6 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt6 import QtCore, QtGui, QtWidgets


class Ui_ScanPort(object):
    def setupUi(self, ScanPort):
        ScanPort.setObjectName("ScanPort")
        ScanPort.resize(800, 600)
        font = QtGui.QFont()
        font.setFamily("JetBrains Mono")
        ScanPort.setFont(font)
        self.centralwidget = QtWidgets.QWidget(ScanPort)
        self.centralwidget.setObjectName("centralwidget")
        self.gridLayout = QtWidgets.QGridLayout(self.centralwidget)
        self.gridLayout.setObjectName("gridLayout")
        self.verticalLayout = QtWidgets.QVBoxLayout()
        self.verticalLayout.setObjectName("verticalLayout")
        self.horizontalLayout_2 = QtWidgets.QHBoxLayout()
        self.horizontalLayout_2.setObjectName("horizontalLayout_2")
        self.label = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(12)
        self.label.setFont(font)
        self.label.setObjectName("label")
        self.horizontalLayout_2.addWidget(self.label)
        self.lineHost = QtWidgets.QLineEdit(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(16)
        self.lineHost.setFont(font)
        self.lineHost.setObjectName("lineHost")
        self.horizontalLayout_2.addWidget(self.lineHost)
        self.line_3 = QtWidgets.QFrame(self.centralwidget)
        self.line_3.setFrameShape(QtWidgets.QFrame.Shape.VLine)
        self.line_3.setFrameShadow(QtWidgets.QFrame.Shadow.Sunken)
        self.line_3.setObjectName("line_3")
        self.horizontalLayout_2.addWidget(self.line_3)
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(12)
        self.label_2.setFont(font)
        self.label_2.setObjectName("label_2")
        self.horizontalLayout_2.addWidget(self.label_2)
        self.linePorts = QtWidgets.QLineEdit(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(16)
        self.linePorts.setFont(font)
        self.linePorts.setObjectName("linePorts")
        self.horizontalLayout_2.addWidget(self.linePorts)
        self.verticalLayout.addLayout(self.horizontalLayout_2)
        self.horizontalLayout_4 = QtWidgets.QHBoxLayout()
        self.horizontalLayout_4.setObjectName("horizontalLayout_4")
        spacerItem = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Policy.Expanding, QtWidgets.QSizePolicy.Policy.Minimum)
        self.horizontalLayout_4.addItem(spacerItem)
        self.radioOrder = QtWidgets.QRadioButton(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(9)
        self.radioOrder.setFont(font)
        self.radioOrder.setChecked(False)
        self.radioOrder.setObjectName("radioOrder")
        self.horizontalLayout_4.addWidget(self.radioOrder)
        self.radioSpeed = QtWidgets.QRadioButton(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(9)
        self.radioSpeed.setFont(font)
        self.radioSpeed.setStatusTip("")
        self.radioSpeed.setChecked(True)
        self.radioSpeed.setObjectName("radioSpeed")
        self.horizontalLayout_4.addWidget(self.radioSpeed)
        spacerItem1 = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Policy.Expanding, QtWidgets.QSizePolicy.Policy.Minimum)
        self.horizontalLayout_4.addItem(spacerItem1)
        self.verticalLayout.addLayout(self.horizontalLayout_4)
        self.buttScan = QtWidgets.QPushButton(self.centralwidget)
        self.buttScan.setEnabled(False)
        self.buttScan.setMinimumSize(QtCore.QSize(0, 40))
        font = QtGui.QFont()
        font.setPointSize(14)
        self.buttScan.setFont(font)
        self.buttScan.setToolTip("")
        self.buttScan.setObjectName("buttScan")
        self.verticalLayout.addWidget(self.buttScan)
        self.line = QtWidgets.QFrame(self.centralwidget)
        self.line.setFrameShape(QtWidgets.QFrame.Shape.HLine)
        self.line.setFrameShadow(QtWidgets.QFrame.Shadow.Sunken)
        self.line.setObjectName("line")
        self.verticalLayout.addWidget(self.line)
        self.textResult = QtWidgets.QTextBrowser(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(14)
        self.textResult.setFont(font)
        self.textResult.setFocusPolicy(QtCore.Qt.FocusPolicy.NoFocus)
        self.textResult.setTabChangesFocus(False)
        self.textResult.setDocumentTitle("")
        self.textResult.setObjectName("textResult")
        self.verticalLayout.addWidget(self.textResult)
        self.horizontalLayout_3 = QtWidgets.QHBoxLayout()
        self.horizontalLayout_3.setObjectName("horizontalLayout_3")
        self.toolClear = QtWidgets.QToolButton(self.centralwidget)
        self.toolClear.setEnabled(False)
        self.toolClear.setMinimumSize(QtCore.QSize(35, 35))
        self.toolClear.setObjectName("toolClear")
        self.horizontalLayout_3.addWidget(self.toolClear)
        spacerItem2 = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Policy.Expanding, QtWidgets.QSizePolicy.Policy.Minimum)
        self.horizontalLayout_3.addItem(spacerItem2)
        self.toolStop = QtWidgets.QToolButton(self.centralwidget)
        self.toolStop.setEnabled(False)
        self.toolStop.setMinimumSize(QtCore.QSize(35, 35))
        self.toolStop.setObjectName("toolStop")
        self.horizontalLayout_3.addWidget(self.toolStop)
        self.verticalLayout.addLayout(self.horizontalLayout_3)
        self.line_2 = QtWidgets.QFrame(self.centralwidget)
        self.line_2.setFrameShape(QtWidgets.QFrame.Shape.HLine)
        self.line_2.setFrameShadow(QtWidgets.QFrame.Shadow.Sunken)
        self.line_2.setObjectName("line_2")
        self.verticalLayout.addWidget(self.line_2)
        self.horizontalLayout = QtWidgets.QHBoxLayout()
        self.horizontalLayout.setObjectName("horizontalLayout")
        spacerItem3 = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Policy.Expanding, QtWidgets.QSizePolicy.Policy.Minimum)
        self.horizontalLayout.addItem(spacerItem3)
        self.toolDelete = QtWidgets.QToolButton(self.centralwidget)
        self.toolDelete.setEnabled(False)
        self.toolDelete.setMinimumSize(QtCore.QSize(75, 25))
        self.toolDelete.setObjectName("toolDelete")
        self.horizontalLayout.addWidget(self.toolDelete)
        self.buttOpen = QtWidgets.QPushButton(self.centralwidget)
        self.buttOpen.setEnabled(False)
        self.buttOpen.setMinimumSize(QtCore.QSize(150, 30))
        font = QtGui.QFont()
        font.setPointSize(12)
        self.buttOpen.setFont(font)
        self.buttOpen.setObjectName("buttOpen")
        self.horizontalLayout.addWidget(self.buttOpen)
        self.buttSave = QtWidgets.QPushButton(self.centralwidget)
        self.buttSave.setEnabled(False)
        self.buttSave.setMinimumSize(QtCore.QSize(150, 30))
        font = QtGui.QFont()
        font.setPointSize(12)
        self.buttSave.setFont(font)
        self.buttSave.setObjectName("buttSave")
        self.horizontalLayout.addWidget(self.buttSave)
        self.verticalLayout.addLayout(self.horizontalLayout)
        self.gridLayout.addLayout(self.verticalLayout, 0, 0, 1, 1)
        ScanPort.setCentralWidget(self.centralwidget)

        self.retranslateUi(ScanPort)
        QtCore.QMetaObject.connectSlotsByName(ScanPort)
        ScanPort.setTabOrder(self.linePorts, self.lineHost)
        ScanPort.setTabOrder(self.lineHost, self.buttScan)
        ScanPort.setTabOrder(self.buttScan, self.buttSave)
        ScanPort.setTabOrder(self.buttSave, self.buttOpen)
        ScanPort.setTabOrder(self.buttOpen, self.textResult)

    def retranslateUi(self, ScanPort):
        _translate = QtCore.QCoreApplication.translate
        ScanPort.setWindowTitle(_translate("ScanPort", "ScanPort"))
        self.label.setText(_translate("ScanPort", "Host:"))
        self.lineHost.setPlaceholderText(_translate("ScanPort", "localhost"))
        self.label_2.setText(_translate("ScanPort", "Port:"))
        self.linePorts.setPlaceholderText(_translate("ScanPort", "* or 1,80,21,1000-1010"))
        self.radioOrder.setToolTip(_translate("ScanPort", "Set if port scan order is important (scan speed is lost on closed ports)"))
        self.radioOrder.setText(_translate("ScanPort", "Order"))
        self.radioSpeed.setToolTip(_translate("ScanPort", "Set if port scan speed is important (scan speed is fast, but open ports jump slightly ahead of their position in the \"result output\" stream due to the implementation of the multithreading scheme)"))
        self.radioSpeed.setText(_translate("ScanPort", "Speed"))
        self.buttScan.setText(_translate("ScanPort", "Scan"))
        self.toolClear.setText(_translate("ScanPort", "Clear"))
        self.toolStop.setText(_translate("ScanPort", "Stop"))
        self.toolDelete.setText(_translate("ScanPort", "Delete save"))
        self.buttOpen.setText(_translate("ScanPort", "Open"))
        self.buttSave.setText(_translate("ScanPort", "Save"))


if __name__ == "__main__":
    import sys
    app = QtWidgets.QApplication(sys.argv)
    ScanPort = QtWidgets.QMainWindow()
    ui = Ui_ScanPort()
    ui.setupUi(ScanPort)
    ScanPort.show()
    sys.exit(app.exec())
